### DISCUSSION TITLE
Error Handling for Data Fetching in Client Components

### DISCUSSION THREAD

**Discussion by SiarheiLazakovich** [vercel/next.js#49204] [Help] [answered]
### Summary

Hey.

The problem is that no way how to handle errors raised after suspended data fetching.
Example is pretty simple and based on "reproduction-template-app-dir" template.
I just added error.tsx and global-error.tsx files and modified page.tsx.

- If you open example you will see that code works as expected.
- If you open page.tsx and uncomment line num.8 you will get error which will be handled with error.tsx, also works as expected.
- But if you uncomment line num.17 in page.tsx you will produce error which will not be handled and you will get nothing.

It possible to handle if add loading.tsx file in the app dir root (I added loading-eg.tsx, just rename it for testing), but in this case SSR returns loader instead wait until data fetched.

### Example

https://codesandbox.io/p/sandbox/focused-wiles-kwrhoq

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I think you have to cache, or at least, return the same promise into the `use` hook:

```tsx
"use client";

import { use, cache } from "react";

const fetchUser = cache(() =>
  fetch("https://randomuser.me/api/").then((data) => data.json())
);

const samePromise = fetch("https://randomuser.me/api/").then((data) =>
  data.json()
);
export default function Home() {
  // raise error
  // will catched with error.tsx
  // throw Error("before");

  const user = use(fetchUser());
  // this one also works
  // const user = use(samePromise);
  // this will crash because every time it is a new promise seen by use
  // const user = use(
  //   fetch("https://randomuser.me/api/").then((data) => data.json())
  // );
  // raise error
  // produce infinite fetch loop
  // how to catch this?
  throw Error("after"); // eg: console.log(user.foo.bar);

  return <pre>{JSON.stringify(user, null, "  ")}</pre>;
}

```

**Reply by SiarheiLazakovich** [reply]
Thank you! You are right, it helps)
